text/0120-js-sdk-scope.md

```diff
@@ -0,0 +1,187 @@
+
+This means that e.g. in Browser JS SDK, all breadcrumbs will be global, while in request-based backend SDKs, they will be per-request.
+
+In cases where you want to isolate breadcrumbs manually (e.g. a queue job), you can manually define a breadcrumbs scope.
+
+## Global Scope
+
+First, there is a _Global Scope_. This scope is tied to the client, and anything added to the global scope will always be added to _all_ events.
+Note that the global scope _is not_ the initial scope, but a separate, "empty" scope. This way, you cannot accidentally pollute it. 
+You can access it via `getGlobalScope()`:
+
+```js
+Sentry.init();
+
+const scope = Sentry.getScope(); // <-- the initial scope, that other scopes will inherit from
+const globalScope = Sentry.getGlobalScope(); // <-- the global scope that will be applied to all events
+
+scope !== globalScope
+```
+
+Global scope data is applied to _all_ events.
+
+### Request Scope
+
+A request scope is a scope valid for a request. This is relevant for Node server SDK usage only.
+Request scope should automatically be set by the SDK. Users can _access_ them in order to add data that should be valid for the whole request, e.g. a user:
+
+```js
+const app = express();
+app.get('/my-route', () => {
+  const requestScope = getRequestScope();
+  requestScope?.setUser(user);
+});
+```
+
+Request scope data is applied to all requests in this request. Request scope is defined to also work e.g. in other callbacks and similar:
+
+```js
+fastify.addHook('preValidation', () => {
+  // You can attach stuff to the request scope even from other places,
+  // ensuring it is added to all events in that request
+  const requestScope = getRequestScope();
+  requestScope.addEventProcessor(...);
+});
+```
+
+Note that like the global scope, the request scope _is not_ the same as the scope in the route handler:
+
+```js
+app.get('/my-route', () => {
+  // anything added here will apply to all events in this request
+  const requestScope = getRequestScope();
+  // anything added here will be inherited normally
+  const scope = getScope();
+  requestScope !== scope;
+});
+```
+
+This is also done to a) avoid accidentally polluting the request scope, and b) to clarify things and avoid relying on internals (e.g. what is the exact scope that is active in a route handler depends on the instrumentation implementation).
+
+If not in a request, `getRequestScope()` will return `undefined`. While this means that users have to always check this return value, it will at least ensure that users capture incorrect cases - otherwise, they may be confused why their code is working but things are not actually added to the request events.
+
+## Breadcrumbs Scope
+
+If you need to manually isolate breadcrumbs, e.g. for a queue function, you can do this with the following new function:
+
+```js
+function myQueueJob() {
+  Sentry.withScope((scope) => {
+    scope.captureBreadcrumbsHere();
+    // do something
+  });
```

Comment:
I'm not sure if I get the Breadcrumbs Scope isn't it just a copy of the initial Scope (in this example.)?

Reply:
Or does this mean that the breadcrumb will be added to the `breadcrumbs scope` of the parent/global?

Reply:
The big difference is that the breadcrumbs scope is applied with the breadcrumbs it has at event creation time, not at the time when the scope is forked. Think of this slightly contrived example, but it should explain the problem:

```js
Sentry.withScope(scope => {
  Sentry.addBreadcrumb({ message: 'breadcrumb 1' }); 
  // scope.breadcrumbs: ['breadcrumb1']

  Sentry.withScope(scope2 => {
    // scope2.breadcrumbs: ['breadcrumb1']
    Sentry.addBreadcrumb({ message: 'breadcrumb2' }); 
    // scope2.breadcrumbs: ['breadcrumb1', 'breadcrumb2']
  });

  // scope.breadcrumbs: ['breadcrumb1']

  Sentry.withScope(scope3 => {
    // scope3.breadcrumbs: ['breadcrumb1']
    Sentry.captureException(error);
  });
});
```

While this appears a bit contrived, in an OTEL world scopes are automatically forked all the time, so you get into this scenario quite fast there. With this proposal, if you define a breadcrumbs scope, instead you'd have this:

```js
Sentry.withScope(scope => {
  scope.captureBreadcrumbsHere();
  Sentry.addBreadcrumb({ message: 'breadcrumb 1' }); 
  // scope.breadcrumbs: []
  // scope.breadcrumbsScope.breadcrumbs: ['breadcrumb1']

  Sentry.withScope(scope2 => {
    // scope2.breadcrumbs: []
    // scope.breadcrumbsScope.breadcrumbs: ['breadcrumb1']
    Sentry.addBreadcrumb({ message: 'breadcrumb2' }); 
    // scope2.breadcrumbs: []
    // scope2.breadcrumbsScope.breadcrumbs: ['breadcrumb1', 'breadcrumb2']
  });

  Sentry.withScope(scope3 => {
    // scope3.breadcrumbs: []
    // scope3.breadcrumbsScope.breadcrumbs: ['breadcrumb1', 'breadcrumb2']
    Sentry.captureException(error);
  });
});
```

Does that make sense? :D

Reply:
Also, in RN you maybe don't need this at all. For Browser JS, most probably just having all breadcrumbs be on the global scope always may be enough. (TBD)

Reply:
Thanks for the explanation. That makes sense. Yeah for RN there won't be much difference from holding the breadcrumbs on the global scope, but I understand the difference. 

The `captureBreadcrumbsHere` enables us to select the scope where from which the `breadcrumbs` will be shared.